lib/internal/vm/localworker.js

```diff
@@ -0,0 +1,139 @@
+// incorporated into Node.js with Anna's permission.
+// See the LICENSE file for LICENSE and copyright attribution.
+
+const {
+  Promise,
+} = primordials;
+
+const {
+  LocalWorker: LocalWorkerImpl,
+} = internalBinding('contextify');
+
+const EventEmitter = require('events');
+const { setTimeout } = require('timers');
+const { dirname, join } = require('path');
+
+let debug = require('internal/util/debuglog').debuglog('localworker', (fn) => {
+  debug = fn;
+});
+
+class LocalWorker extends EventEmitter {
+  #handle = undefined;
+  #process = undefined;
+  #global = undefined;
+  #module = undefined;
+  #stoppedPromise = undefined;
+
+  /**
+   */
+  constructor() {
+    super();
+    this.#handle = new LocalWorkerImpl();
+    this.#handle.onexit = (code) => {
+      this.stop();
+      this.emit('exit', code);
+    };
+    try {
+      this.#handle.start();
+      this.#handle.load((process, nativeRequire, globalThis) => {
+        this.#process = process;
+        this.#module = nativeRequire('module');
+        this.#global = globalThis;
+        process.on('uncaughtException', (err) => {
+          if (process.listenerCount('uncaughtException') === 1) {
+            // If we are stopping, silence all errors
+            if (!this.#stoppedPromise) {
+              this.emit('error', err);
+            }
+            process.exit(1);
+          }
+        });
+      });
+    } catch (err) {
+      this.#handle.stop();
+      throw err;
+    }
+  }
+
+  /**
+   * @returns {Promise<void>}
+   */
+  async stop() {
+    // TODO(@mcollina): add support for AbortController, we want to abort this,
+    // or add a timeout.
+    return this.#stoppedPromise ??= new Promise((resolve) => {
+      const tryClosing = () => {
+        const closed = this.#handle.tryCloseAllHandles();
+        debug('closed %d handles', closed);
+        if (closed > 0) {
+          // This is an active wait for the handles to close.
+          // We might want to change this in the future to use a callback,
+          // but at this point it seems like a premature optimization.
+          // TODO(@mcollina): refactor to use a close callback
+          setTimeout(tryClosing, 100);
+        } else {
+
+          this.#handle.stop();
+          resolve();
+        }
+      };
+
+      // We use setTimeout instead of setImmediate because it runs in a different
+      // phase of the event loop. This is important because the immediate queue
+      // would crash if the environment it refers to has been already closed.
+      setTimeout(tryClosing, 100);
+    });
+  }
+
+  get process() {
+    return this.#process;
+  }
+
+  get globalThis() {
+    return this.#global;
+  }
+
+  createRequire(...args) {
+    return this.#module.createRequire(...args);
+  }
+
+  /**
+   * @param {() => any} method
+   */
+  runInWorkerScope(method) {
+    return this.#handle.runInCallbackScope(method);
+  }
+
+  /**
+   * @param {string} filename
+   */
+  async createImport(filename) {
```

Comment:
Why is it called `createImport`, instead of just `import`?

Also my guess is this might be what you are looking for:

```js
this.#loader = require('internal/process/esm_loader').esmLoader;
this.#loader.import(specifiers, parentURL, importAssertions);
```

Because:

1. `ShadowRealm#importValue` basically just invokes the host dynamically import callback to get the exports
2. This is what our host dynamically import callback for CommonJS modules essentially do.

Reply:
Another difference with `ShadowRealm.prototype.importValue` is that ShadowRealm's module resolution has no `parentURL` -- in the browser, the specifier is resolved relative to the document's origin, like <script> tags. 

If we create the `LocalWorker`/`NodeRealm` with an origin (or its working directory) set with the constructor, we can have similar method signatures to the dynamic `import` call:

```typescript
interface LocalWorker {
  constructor(cwd: string);
  require(specifier): any;
  import(specifier, assertions): ModuleNamespace;
}
```

Reply:
I'll try interacting with the ESM loader directly

Reply:
```js
this.#loader = require('internal/process/esm_loader').esmLoader;
this.#loader.import(specifiers, parentURL, importAssertions);
```

@joyeecheung That would load the `esmLoader` of the external environment. I don't know how to run that code in the inner context/environment. In other terms, I need to grab a `require` implementation from the inner context/environment that can load from `internal/`.

Reply:
You could get the internal require of the new context from C++ land via something like `Realm::GetCurrent(new_context)->builtin_module_require()`

Reply:
Thanks, this is super handy! I'll try it.

Reply:
It works. Thanks @joyeecheung for the support!